application/src/main/java/org/togetherjava/tjbot/features/javamail/RSSHandlerRoutine.java

```diff
@@ -0,0 +1,417 @@
+        this.database = database;
+        this.fallbackChannelPattern =
+                Pattern.compile(this.config.fallbackChannelPattern()).asMatchPredicate();
+        this.targetChannelPatterns = new HashMap<>();
+        this.config.feeds().forEach(feed -> {
+            if (feed.targetChannelPattern() != null) {
+                Predicate<String> predicate =
+                        Pattern.compile(feed.targetChannelPattern()).asMatchPredicate();
+                targetChannelPatterns.put(feed, predicate);
+            }
+        });
+        this.rssReader = new RssReader();
+    }
+
+    @Override
+    @NotNull
+    public Schedule createSchedule() {
+        return new Schedule(ScheduleMode.FIXED_DELAY, 0, interval, TimeUnit.MINUTES);
+    }
+
+    @Override
+    public void runRoutine(@Nonnull JDA jda) {
+        this.config.feeds().forEach(feed -> sendRSS(jda, feed));
+    }
+
+    /**
+     * Sends all the necessary posts from a given RSS feed.
+     * <p>
+     * This handles fetching the latest posts from the given URL, checking which ones have already
+     * been posted by reading information from the database and updating the last posted date.
+     *
+     * @param jda The JDA instance.
+     * @param feedConfig The configuration object for the RSS feed.
+     */
+    private void sendRSS(JDA jda, RSSFeed feedConfig) {
+        List<TextChannel> textChannels = getTextChannelsFromFeed(jda, feedConfig);
+        if (textChannels.isEmpty()) {
+            logger.warn("Tried to send an RSS post, got empty response (channel {} not found)",
+                    feedConfig.targetChannelPattern());
+            return;
+        }
+
+        List<Item> rssItems = fetchRSSItemsFromURL(feedConfig.url());
+        if (rssItems.isEmpty()) {
+            return;
+        }
+
+        for (Item item : rssItems) {
+            if (!isValidDateFormat(item, feedConfig)) {
+                logger.warn("Could not find valid or matching date format for RSS feed {}",
+                        feedConfig.url());
+                return;
+            }
+        }
+
+        final Predicate<Item> shouldItemBePosted = prepareItemPostPredicate(feedConfig, rssItems);
+        if (shouldItemBePosted == null) return;
+        rssItems.reversed()
+            .stream()
+            .filter(shouldItemBePosted)
+            .forEachOrdered(item -> postItem(textChannels, item, feedConfig));
+    }
+
+    private Predicate<Item> prepareItemPostPredicate(RSSFeed feedConfig, List<Item> rssItems) {
+        Optional<RssFeedRecord> rssFeedRecord = getRssFeedRecordFromDatabase(feedConfig);
+        Optional<ZonedDateTime> lastPostedDate =
+                getLatestPostDateFromItems(rssItems, feedConfig.dateFormatterPattern());
+
+        lastPostedDate.ifPresent(
+                date -> updateLastDateToDatabase(feedConfig, rssFeedRecord.orElse(null), date));
+
+        if (rssFeedRecord.isEmpty()) {
+            return null;
```

Comment:
You should avoid returning null and instead use Optional or throw exception depending on context

Reply:
Done!